core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -102,39 +102,40 @@ final class FiberRefs private (
                 childStack: List[(FiberId.Runtime, Any)],
                 childDepth: Int,
                 childModified: Boolean = false
-              ): (Any, Boolean) =
+              ): (Any, Any, Boolean) =
                 (parentStack, childStack) match {
-                  case ((parentFiberId, _) :: parentAncestors, (childFiberId, childValue) :: childAncestors) =>
+                  case (
+                        (parentFiberId, parentValue) :: parentAncestors,
+                        (childFiberId, childValue) :: childAncestors
+                      ) =>
                     if (parentFiberId == childFiberId)
-                      (childValue, childModified)
+                      (parentValue, childValue, childModified)
                     else if (childDepth > parentDepth)
                       findAncestor(parentStack, parentDepth, childAncestors, childDepth - 1, true)
                     else if (childDepth < parentDepth)
                       findAncestor(parentAncestors, parentDepth - 1, childStack, childDepth, childModified)
                     else
                       findAncestor(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1, true)
                   case _ =>
-                    (ref.initial, true)
+                    (ref.initial, ref.initial, true)
                 }
 
-              val (ancestor, wasModified) = findAncestor(parentStack, parentDepth, childStack, childDepth)
+              val (parentAncestor, childAncestor, wasModified) =
+                findAncestor(parentStack, parentDepth, childStack, childDepth)
 
-              if (!wasModified) parentFiberRefs
+              val patch = if (wasModified) ref.diff(childAncestor, childValue) else ref.diff(parentAncestor, childValue)
```

Comment:
I don't believe this condition is the right way to determine whether we are joining our parent or not.

Consider this example (which is still failing on this branch):
```scala
for {
  prom <- Promise.make[Nothing, Fiber.Runtime[Nothing, Unit]]
  f <- prom.await.flatMap(_.inheritAll).fork
  _ <- ZIO.scoped {
        ZIO.scoped {
          ZIO.unit.fork.flatMap(prom.succeed)
        }.fork
      }
  _ <- f.join
} yield ()
```

This will enter the joinAs function with the following ancestries:
Parent: 
```scala
Map(zio.FiberRef$unsafe$$anon$2@28999e37 -> (List((Runtime(76,1692627579297,zio.FooSpec.spec(FooSpec.scala:11)),ZEnvironment())),1))
```
Child:
```scala
Map(zio.FiberRef$unsafe$$anon$2@28999e37 -> (List((Runtime(78,1692627579298,zio.FooSpec.spec(FooSpec.scala:68)),ZEnvironment(Scope -> zio.Scope$$anon$2@2745c2ef)), (Runtime(76,1692627579297,zio.FooSpec.spec(FooSpec.scala:11)),ZEnvironment(Scope -> zio.Scope$$anon$2@3c206b9f))),2))
```

We have an update in the child after the fork point, so `wasModified` will be true -> this heuristic won't work in this case.
---
More generally, I think the problem here is that the two fibers have a different view of what the value of the fiberref was before forking (parent thinks it's an empy ZEnvironment, child thinks it has a scope in there.) and we can't really determine which one is correct.

IMO, the easiest way to solve this is probably to include a global / per fiberref counter in the values we store in the fiberRefLocals so we can determine and pick the logically earlier value when the two fibers don't agree on the value for a particular fiberref (like above).

Reply:
@mschuwalow Will take a look!

Reply:
Yes I I don't think your example is actually about a child joining a parent but a fiber joining its "nephew".

It can be generalized to this. If we have a fiber `parent` that forks `child1`, makes some change, and then forks `child2`, what should be the behavior if `child1` joins `child2`, or vice versa?

Currently we compute the patch based on the difference between the joined fiber's ancestor value and its current value. So if `child2` joins `child1` it correctly incorporates only the changes made by `child1`. However, if `child1` joins `child2`, `child1` will only see the changes made by `child2` and not the change made by `parent` after forking `child1` but before forking `child2`.

Likewise, if we computed the patch based on the difference between the joining fiber's ancestor value and the joined fiber's current value, we would be incorrect in some cases. If `child1` joins `child2` we now correctly pick up all the changes. But if `child2` joins `child1` it would compute the difference between its ancestor value, including the update made by `parent`, and the value of `child1`, incorrectly concluding that the change made by `parent` had been removed by `child1`.

As you say, we need a concept of a "version" of the value of a `FiberRef` for a fiber so we can avoid holding on to all of the previous `FiberRef` values but know when we join a fiber whether differences in the `FiberRef` value for our common ancestor were made after we were forked, and then should be incorporated by us, or before we were forked, and thus we should ignore.

Made these changes in the latest version.

Reply:
great, many thanks!

Reply:
Thank you! 🙏